Sources/BSON/BSONCode.swift

```diff
@@ -0,0 +1,82 @@
+import NIO
+
+/// A struct to represent the BSON Code type.
+public struct BSONCode: Equatable, Hashable {
+    /// A string containing Javascript code.
+    public let code: String
+
+    /// Initializes a `BSONCode` with an optional scope value.
+    public init(code: String) {
+        self.code = code
+    }
+}
+
+/// A struct to represent BSON CodeWithScope.
+public struct BSONCodeWithScope: Equatable, Hashable {
+    /// A string containing Javascript code.
+    public let code: String
+
+    /// An optional scope `BSONDocument` containing a mapping of identifiers to values,
+    /// representing the context in which `code` should be evaluated.
+    public let scope: BSONDocument
+
+    /// Initializes a `BSONCodeWithScope` with an optional scope value.
+    public init(code: String, scope: BSONDocument) {
+        self.code = code
+        self.scope = scope
+    }
+}
+
+extension BSONCode: BSONValue {
+    internal static var bsonType: BSONType { .code }
+
+    internal var bson: BSON { .code(self) }
+
+    internal static func read(from buffer: inout ByteBuffer) throws -> BSON {
+        guard let code = try String.read(from: &buffer).stringValue else {
+            throw BSONError.InternalError(message: "Cannot code")
+        }
+        return .code(BSONCode(code: code))
+    }
+
+    internal func write(to buffer: inout ByteBuffer) {
+        self.code.write(to: &buffer)
+    }
+}
+
+extension BSONCodeWithScope: BSONValue {
+    internal static var bsonType: BSONType { .codeWithScope }
+
+    internal var bson: BSON { .codeWithScope(self) }
+
+    internal static func read(from buffer: inout ByteBuffer) throws -> BSON {
+        let reader = buffer.readerIndex
+        guard let size = buffer.readInteger(endianness: .little, as: Int32.self) else {
+            throw BSONError.InternalError(message: "Cannot code with scope size")
+        }
+        guard size >= 14, (size - 4) < buffer.readableBytes else {
+            throw BSONError.InternalError(message: "Code with scope has size: \(size) but there "
```

Comment:
this error message only applies to the latter condition in this `guard`, so it needs to be split up so each condition can have an appropriate error message.

also, including a comment indicating how the 14 number was calculated would help future readers.